app/queries/json_api.rb

```diff
@@ -1,27 +1,73 @@
 class JsonApi
   class Error < StandardError; end
+
   class InvalidApiRootError < Error; end
   class MissingPrivateKeyError < Error; end
-  class InvalidResponseError < Error; end
+
+  class ApiError < Error; end
+  class InvalidResponseError < ApiError; end
+  class ConnectionError < ApiError; end
+  class StatusBadRequestError < ApiError; end
+  class StatusNotFoundError < ApiError; end
+  class StatusServerError < ApiError; end
+  class StatusUnexpectedError < ApiError; end
 
   def initialize(api_config = {})
     @connection = ConnectionBuilder.new.build(api_config)
   end
 
   def get(path)
-    @connection.get(path).body
-  rescue Faraday::ParsingError => e
-    raise InvalidResponseError, e.message
+    response = call_and_raise do
+      @connection.get(path)
+    end
+    response.body
   end
 
   def post(path, params)
-    response = @connection.post(path) do |request|
-      request.body = params.to_json
-      request.headers['Content-Type'] = 'application/json'
+    response = call_and_raise do
+      @connection.post(path) do |request|
+        request.body = params.to_json
+        request.headers['Content-Type'] = 'application/json'
+      end
     end
     response.body
+  end
+
+  private
+
+  def call_and_raise
+    yield.tap { |response| raise_for_http_status(response) }
   rescue Faraday::ParsingError => e
     raise InvalidResponseError, e.message
+  rescue Faraday::ConnectionFailed => e
+    raise ConnectionError, e
```

Comment:
Why is this exception re-raised with all the exception metadata, whereas the above exception "InvalidResponseError" is re-raised passing only `e.message`.

Reply:
Good spot - this should be made consistent.

It actually has the same result: the second argument to raise is a string, so it casts the error to a string, which returns the same value as e.message.

Reply:
Amended the commit